pkg/ngrokgodriver/driver.go

```diff
@@ -0,0 +1,121 @@
+package ngrokgodriver
+
+import (
+	"context"
+	"io"
+	"log"
+	"net"
+	"reflect"
+
+	"golang.ngrok.com/ngrok"
+	"golang.ngrok.com/ngrok/config"
+	"golang.org/x/sync/errgroup"
+)
+
+type TunnelManager struct {
+	session ngrok.Session
+	tunnels map[string]ngrok.Tunnel
+}
+
+func NewTunnelManager() (*TunnelManager, error) {
+	opts := []ngrok.ConnectOption{
+		ngrok.WithAuthtokenFromEnv(),
+	}
+	session, err := ngrok.Connect(context.Background(), opts...)
+	if err != nil {
+		return nil, err
+	}
+	return &TunnelManager{
+		session: session,
+		tunnels: make(map[string]ngrok.Tunnel),
+	}, nil
+}
+
+func (tm *TunnelManager) CreateTunnel(ctx context.Context, t TunnelsAPIBody) error {
+	if tun, ok := tm.tunnels[t.Name]; ok {
+		if reflect.DeepEqual(tun.Labels(), t.Labels) {
+			// The tunnel already exists and has the same labels, do nothing
+			return nil
+		}
+		// There is already a tunnel with this name, start the new one and defer closing the old one
+		defer tm.stopTunnel(context.Background(), tun)
+	}
+
+	tun, err := tm.session.Listen(ctx, tm.buildTunnelConfig(t))
+	if err != nil {
+		return err
+	}
+	tm.tunnels[t.Name] = tun
+	go tm.startTunnel(tun, t.Addr)
+	return nil
+}
+
+func (tm *TunnelManager) DeleteTunnel(ctx context.Context, name string) error {
+	tun := tm.tunnels[name]
+	if tun == nil {
+		return nil
```

Comment:
This makes sense within the context of this function. If the tun is nil, there isn't anything left to do. For the controller calling this though, I wonder if there is a code path where this may happen and be valid. My assumption is that if things are all working as they are supposed to, we shouldn't hit this. And if we are, then maybe we are getting extra reconcile loops or something. Do you think hitting this case deserves a debug/info log so we might happen to see if we are hitting this case, or do you think those wouldn't be helpful and we should look at something more like tracing?

Reply:
I'd go with logging initially. Its hard to ask people to give traces in issue reports and requires them to have tracing infrastructure set up. If that sounds good, I'll add a log statement to this and give the tunnel manager an actual logger.

Reply:
yep that makes sense on the tracing stuff. thanks!

Reply:
Fixed in https://github.com/ngrok/ngrok-ingress-controller/pull/85/commits/d320396e9a92e568b567762b17ea16021a1927b2